### PR TITLE
Foxtrot Rework + Marine Reinforcements

### PR DESCRIPTION
**Pull request by zc-byte-boxer**
# Marine Reinforcement + Foxtrot Rework

This is a PR that rebalance Foxtrot, making it a one time use, at a higher price, but giving more marines with a pop-scale mechanic.

As well, this adds at the nuke tier an option to call a nerfed ERT team for additional reinforcements.
# Explain why it's good for the game

https://hackmd.io/@byte-boxer/BypH9VZd0

Outside of the design doc, we'vemade our rounds longer over the last year, but have progressively nerfed marine respawns. This PR adds more opportunities for the marines to respawn, while ridding the marines of the option to buy more powerful reinforcements. 

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Tested on local. Can manually call ERTs, but everything isn't working yet.

</details>

# Changelog
:cl: Byte-Boxer
add: Adds Beacon Reinforcements at Tier Four to the techweb. It allows marines to call, at the price of ten points, for nerfed versions of the UPP, VAIPO, Freelancer, and RMC ERT. One time use only. 
del: Specialist was removed as a techweb buy, though kept in code incase admins wanted to spawn it.
balance: Foxtrot has gone through a major balance overhaul. It now is a one time use only, and costs more, but has a minimum spawn of 6, and at 100, 110, 120 pop, it gains 2 more members before maxing out at 12. Foxtrot no longer has a side spec that can be bought, but does have an SG.
/:cl:

# Current Bugs
- Beacon-Reinforcements is in the techweb code folder, but does not actually appear visible clientside when looking at techwebs.
- These special ERTs need to have no minimum number to spawn, or need to relaunch if they fail to get anyone.
- Spawning these ERTs as freed mobs if they're not all picked doesn't work currently.
- Linters failure.

### CODE REVIEW

File: code/datums/emergency_calls/beacon-reinforcements/freelancers.dm [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		to_chat(mob, SPAN_ROLE_HEADER("You are an independent medic, operating as a mercenary for hire."))
+		arm_equipment(mob, /datum/equipment_preset/other/freelancer/medic, TRUE, TRUE)
+	else if(spawning_mind)
+		to_chat(mob, SPAN_ROLE_HEADER("You are an independent mercenary, operating as a gun for hire."))
+		arm_equipment(mob, /datum/equipment_preset/other/freelancer/beacon, TRUE, TRUE)
+
+	print_backstory(mob)
+
+	sleep(10)
+	if(!mind)
```
Comment (ihatethisengine) [inline code/datums/emergency_calls/beacon-reinforcements/freelancers.dm:45]:
```suggestion
	if(!spawning_mind)
```

File: code/datums/emergency_calls/beacon-reinforcements/freelancers.dm [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/datum/emergency_call/freelancer/beacon/create_member(datum/mind/spawning_mind, turf/override_spawn_loc)
+	var/turf/spawn_loc = override_spawn_loc ? override_spawn_loc : get_spawn_point()
+
+	if(!istype(spawn_loc))
+		return //Didn't find a useable spawn point.
+
+	var/mob/living/carbon/human/mob = new(spawn_loc)
+	spawning_mind.transfer_to(mob, TRUE)
+
+	if(!leader && HAS_FLAG(mob.client.prefs.toggles_ert, PLAY_LEADER) && check_timelock(mob.client, JOB_SQUAD_LEADER, time_required_for_job && (spawning_mind)))
```
Comment (ihatethisengine) [inline code/datums/emergency_calls/beacon-reinforcements/freelancers.dm:30]:
```suggestion
	if(!leader && (!spawning_mind || (HAS_FLAG(mob.client.prefs.toggles_ert, PLAY_LEADER) && check_timelock(mob.client, JOB_SQUAD_LEADER, time_required_for_job))))
```
and implement this for all the similar occurrences. Basically if you don't have a mind (a client willing to roll the role), you have no one to check for preferences.

File: code/datums/emergency_calls/beacon-reinforcements/rmc.dm [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+/obj/effect/landmark/ert_spawns/distress_twe
+	name = "Distress_TWE"
+
+/obj/effect/landmark/ert_spawns/distress_twe/item
+	name = "Distress_TWEItem"
+
+
+	sleep(10)
+	if(!mind)
```
Comment (ihatethisengine) [inline code/datums/emergency_calls/beacon-reinforcements/rmc.dm:55]: wrong place, should be in create_member

**Comment by ihatethisengine** [comment]
Didnt look through everything, but this should be good enough to start with

File: code/datums/emergency_calls/cryo_marines.dm [hunk 4]
```diff
@@ -95,9 +95,33 @@
 /obj/effect/landmark/ert_spawns/distress_cryo
 	name = "Distress_Cryo"
 
-/datum/emergency_call/cryo_squad/tech
-	name = "Marine Cryo Reinforcements (Tech)"
-	mob_max = 5
-	max_engineers = 1
-	max_medics = 1
-	max_heavies = 0
+/proc/get_marines()
```
Comment (kiVts) [inline code/datums/emergency_calls/cryo_marines.dm:98]: Highly suggesting to make a proper path for this. Right now anything and anywhere in the codebase can call this, does it need to be like that?  
e.g. 
`/datum/emergency_call/cryo_squad/tech/proc/get_marines()`
Also move it down below the `/datum/emergency_call/cryo_squad/tech/` after the path is given.

File: code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements
Comment (kiVts) [inline code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements:1]: add .dm at the end. its just a textfile now.

File: code/datums/emergency_calls/beacon-reinforcements/rmc.dm [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	mob_min = 1
+	spawn_max_amount = TRUE
+	home_base = /datum/lazy_template/ert/twe_station
+	shuttle_id = MOBILE_SHUTTLE_ID_ERT4
+	name_of_spawn = /obj/effect/landmark/ert_spawns/distress_twe/beacon
+	item_spawn = /obj/effect/landmark/ert_spawns/distress_twe/item
+
+/datum/emergency_call/royal_marines/beacon/New()
+	..()
+	arrival_message = "[MAIN_SHIP_NAME], this is the HMS Kurtz of the Three World Empires. We've received your call and are enroute to aid."
```
Comment (kiVts) [inline code/datums/emergency_calls/beacon-reinforcements/rmc.dm:13]: Kurtz is a CBRN ship. Thunderchild is the HMS ship used in cm. You cal call it whatever though.
```suggestion
	arrival_message = "[MAIN_SHIP_NAME], this is the HMS Thunderchild of the Three World Empires. We've received your call and are enroute to aid."
```

File: code/datums/emergency_calls/cryo_marines.dm [hunk 6]
```diff
@@ -95,9 +95,33 @@
...
+    var/pop_scale = 0
+    var/count = 0
+
+    for (var/mob/living/carbon/human/H in GLOB.human_mob_list)
```
Comment (kiVts) [inline code/datums/emergency_calls/cryo_marines.dm:102]:
```suggestion
    for (var/mob/living/carbon/human/H in GLOB.alive_human_list)
```

All humans completely or just alive ones?

File: code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements
Comment (kiVts) [reply code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements:1]: also swap the - with _

Comment (kiVts) [reply code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements:1]: same for the folder you created

File: code/datums/emergency_calls/cryo_marines.dm (hunk 6 above)
Comment (Diegoflores31) [reply code/datums/emergency_calls/cryo_marines.dm:102]: it says that it scales with POP not with alive marines tho

Comment (kiVts) [reply code/datums/emergency_calls/cryo_marines.dm:102]: already scales with humans either way /shrug

File: code/datums/emergency_calls/beacon-reinforcements/freelancers.dm (hunk 1 above)
Comment (zc-byte-boxer) [reply code/datums/emergency_calls/beacon-reinforcements/freelancers.dm:50]: changed

File: code/datums/emergency_calls/beacon-reinforcements/freelancers.dm (hunk 2 above)
Comment (zc-byte-boxer) [reply code/datums/emergency_calls/beacon-reinforcements/freelancers.dm]: changed

File: code/datums/emergency_calls/beacon-reinforcements/rmc.dm (hunk 3 above)
Comment (zc-byte-boxer) [reply code/datums/emergency_calls/beacon-reinforcements/rmc.dm:46]: changed

File: code/datums/emergency_calls/beacon-reinforcements/rmc.dm (hunk 5 above)
Comment (zc-byte-boxer) [reply code/datums/emergency_calls/beacon-reinforcements/rmc.dm]: changed it up to the HMS Thanes. I want to keep it as a different name than the other two ERT ships so I don't have to add anything to check if a ship name has been taken

File: code/datums/emergency_calls/cryo_marines.dm (hunk 6 above)
Comment (zc-byte-boxer) [reply code/datums/emergency_calls/cryo_marines.dm:102]: ideally, it should scale off the whole server pop, but I'm not sure how to rewrite it for that

File: code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements
Comment (zc-byte-boxer) [reply code/modules/cm_tech/techs/marine/tier4/beacon-reinforcements:1]: added .dm and changed it up. anything else?

**Comment by zc-byte-boxer** [comment]
https://streamable.com/cy8zu9
[Streamable](https://streamable.com/)

File: code/datums/emergency_calls/cryo_marines.dm [hunk 7]
```diff
@@ -92,12 +92,36 @@
...
-	name = "Marine Cryo Reinforcements (Tech)"
-	mob_max = 5
-	max_engineers = 1
-	max_medics = 1
-	max_heavies = 0
+/datum/emergency_call/cryo_squad/tech/proc/get_marines()
+	var/pop_scale = 0
+	var/count = 0
+
+	for (var/mob/living/carbon/human/H in GLOB.human_mob_list)
```
Comment (kiVts) [inline code/datums/emergency_calls/cryo_marines.dm:110]: if you want it to scale with server pop, use GLOB.clients, dont use for loop, just single var with length(GLOB.clients)

File: code/modules/cm_tech/techs/marine/tier3/cryorine.dm [hunk 8]
```diff
@@ -1,25 +1,16 @@
...
 	tier = /datum/tier/three
 
-/datum/tech/repeatable/cryomarine/can_unlock(mob/M)
-	. = ..()
-	if(!.)
-		return
-	if(!SSticker.mode)
-		to_chat(M, SPAN_WARNING("You can't do this right now!"))
-		return
-
```
Comment (kiVts) [inline code/modules/cm_tech/techs/marine/tier3/cryorine.dm:22]: Why did you remove this

File: code/datums/emergency_calls/beacon-reinforcements/vaipo.dm [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	objectives = "Assist the USCM forces"
+	arrival_message = "[MAIN_SHIP_NAME], this is USCSS Verloc with Vanguard's Arrow Incorporated, Primary Operations; we read your signal, leathernecks, and are coming to aid in accordance with the Military Aid Act of 2177."
+
+/datum/emergency_call/vaipo/beacon/create_member(datum/mind/mind, turf/override_spawn_loc)
+	set waitfor = 0
+	if(SSmapping.configs[GROUND_MAP].map_name == MAP_WHISKEY_OUTPOST)
+		name_of_spawn = /obj/effect/landmark/ert_spawns/distress_wo
+	var/turf/spawn_loc = override_spawn_loc ? override_spawn_loc : get_spawn_point()
+
+	if(!istype(spawn_loc)) return //Didn't find a useable spawn point.
```
Comment (kiVts) [inline code/datums/emergency_calls/beacon-reinforcements/vaipo.dm:22]: no inline ifs

File: code/datums/emergency_calls/beacon-reinforcements/upp.dm [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	objectives = "Assist the USCM forces"
+	arrival_message = "[MAIN_SHIP_NAME], this is the SSV Haldin of the Union of Progressive Peoples. We hear your call for reinforcements and are sending our forces to assist you under International Law as outlined by the Treaty of Canton."
+
+/datum/emergency_call/upp/beacon/create_member(datum/mind/mind, turf/override_spawn_loc)
+	set waitfor = 0
+	if(SSmapping.configs[GROUND_MAP].map_name == MAP_WHISKEY_OUTPOST)
+		name_of_spawn = /obj/effect/landmark/ert_spawns/distress_wo
+	var/turf/spawn_loc = override_spawn_loc ? override_spawn_loc : get_spawn_point()
+
+	if(!istype(spawn_loc)) return //Didn't find a useable spawn point.
```
Comment (kiVts) [inline code/datums/emergency_calls/beacon-reinforcements/upp.dm:25]: no inline ifs, I know this is copypasted, but that code is also six years old.

File: code/datums/emergency_calls/beacon-reinforcements/rmc.dm [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	objectives = "Assist the USCM forces"
+	arrival_message = "[MAIN_SHIP_NAME], this is the HMS Thames of the Three World Empires. We've received your call and are enroute to aid per the C2 Collaborative Defense Agreement."
+
+/datum/emergency_call/rmc/beacon/create_member(datum/mind/mind, turf/override_spawn_loc)
+	set waitfor = 0
+	if(SSmapping.configs[GROUND_MAP].map_name == MAP_WHISKEY_OUTPOST)
+		name_of_spawn = /obj/effect/landmark/ert_spawns/distress_wo
+	var/turf/spawn_loc = override_spawn_loc ? override_spawn_loc : get_spawn_point()
+
+	if(!istype(spawn_loc)) return //Didn't find a useable spawn point.
```
Comment (kiVts) [inline code/datums/emergency_calls/beacon-reinforcements/rmc.dm:22]: Same

File: code/datums/emergency_calls/beacon-reinforcements/freelancers.dm [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+	objectives = "Assist the USCM forces"
+	arrival_message = "[MAIN_SHIP_NAME], this is the Free Vessel Nellie. We've accepted your contract for aid and are coming aboard. Jolly ho!"
+
+/datum/emergency_call/freelancer/beacon/create_member(datum/mind/mind, turf/override_spawn_loc)
+	set waitfor = 0
+	if(SSmapping.configs[GROUND_MAP].map_name == MAP_WHISKEY_OUTPOST)
+		name_of_spawn = /obj/effect/landmark/ert_spawns/distress_wo
+	var/turf/spawn_loc = override_spawn_loc ? override_spawn_loc : get_spawn_point()
+
+	if(!istype(spawn_loc)) return //Didn't find a useable spawn point.
```
Comment (kiVts) [inline code/datums/emergency_calls/beacon-reinforcements/freelancers.dm:22]: same

**Review by kiVts — CHANGES_REQUESTED** [review]
rename this folder, replace the - with _. Adjust accordingly in dme file.
image
Otherwise I probably missed something

File: code/datums/emergency_calls/beacon-reinforcements/rmc.dm (hunk 5 above)
Comment (Steelpoint) [reply code/datums/emergency_calls/beacon-reinforcements/rmc.dm]: The USS Kurtz map can be used for any USCM faction ERT, not just CBRN.

**Comment by harryob** [comment]
pr seems to be stale, let us know if you want it re-opening, or just make a fresh pr for it
